libsmtutil/CHCSmtLib2Interface.cpp

```diff
@@ -223,3 +225,289 @@ std::string CHCSmtLib2Interface::createHeaderAndDeclarations() {
 std::string CHCSmtLib2Interface::createQueryAssertion(std::string name) {
 	return "(assert\n(forall " + forall() + "\n" + "(=> " + name + " false)))";
 }
+
+class SMTLibTranslationContext
+	{
+		CHCSmtLib2Interface const& m_chcInterface;
+		std::map<std::string, SortPointer> knownVariables;
+
+		bool isNumber(std::string const& _expr)
+		{
+			for (char c: _expr)
+				if (!isDigit(c) && c != '.')
+					return false;
+			return true;
+		}
+
+
+	public:
+		explicit SMTLibTranslationContext(CHCSmtLib2Interface const& _chcInterface) : m_chcInterface(_chcInterface)
+		{
+			// fill user defined sorts and constructors
+			auto const& userSorts = m_chcInterface.smtlib2Interface()->userSorts();
+			for (auto const& declaration: userSorts | ranges::views::values)
+			{
+				std::stringstream ss(declaration);
+				SMTLib2Parser parser(ss);
+				auto expr = parser.parseExpression();
+				solAssert(parser.isEOF());
+				solAssert(!isAtom(expr));
+				auto const& args = asSubExpressions(expr);
+				solAssert(args.size() == 3);
+				solAssert(isAtom(args[0]) && asAtom(args[0]) == "declare-datatypes");
+				// args[1] is the name of the type
+				// args[2] is the constructor with the members
+				solAssert(!isAtom(args[2]) && asSubExpressions(args[2]).size() == 1 && !isAtom(asSubExpressions(args[2])[0]));
+				auto const& constructors = asSubExpressions(asSubExpressions(args[2])[0]);
+				solAssert(constructors.size() == 1);
+				auto const& constructor = constructors[0];
+				// constructor is a list: name + members
+				solAssert(!isAtom(constructor));
+				auto const& constructorArgs = asSubExpressions(constructor);
+				for (unsigned i = 1u; i < constructorArgs.size(); ++i)
+				{
+					auto const& carg = constructorArgs[i];
+					solAssert(!isAtom(carg) && asSubExpressions(carg).size() == 2);
+					auto const& nameSortPair = asSubExpressions(carg);
+					solAssert(isAtom(nameSortPair[0]));
+					knownVariables.insert({asAtom(nameSortPair[0]), toSort(nameSortPair[1])});
```

Comment:
```suggestion
					knownVariables.emplace(asAtom(nameSortPair[0]), toSort(nameSortPair[1]));
```

You can apply this for all of the `knownVariables.insert(...)` occurrences.

Reply:
Fixed!